### ISSUE TITLE
Custom error message for missing `end`

### ISSUE THREAD

**Issue by kostmo** [swarm-game/swarm#1141]
I've found a common mistake in my own swarm-lang code to be a missing `end` keyword at end of a `def`.  One reason for the frequency of this error is probably that the languages I, and I suspect most others, are most familiar with (e.g. Python and Haskell) do not require this keyword at the end of a function.

The error manifests itself far away (at the end of the file) and as the message `"unexpected end of input"`.
Since it is a common error, could we special-case this error to have a more user-friendly message?  Perhaps in the parser we can count to make sure that the number of `end` keywords matches the number of `def` keywords.

**Comment by byorgey** [comment]
Yes, that sounds like it should be feasible.  If we work a little harder we can probably even pinpoint the place where the missing `end` likely occurred --- it will be the first place we try to parse a `def` inside the body of another `def`.

**Comment by ussgarci** [comment]
I'd like to work on this issue.  :sweat_smile:

**Comment by xsebek** [comment]
Just to update, since #1087 the `def` inside `def` is valid:
```Haskell
def a =
  def b = 1 end;
  return (b + b)
end;
a
```

It's still a good bet that the `end` should be before the next `def` though.  🙂
